File: csharp-selenium-webdriver-sample/WebDriverFactory.cs
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using OpenQA.Selenium;
using OpenQA.Selenium.Chrome;
using OpenQA.Selenium.Firefox;
using System;

namespace CSharpSeleniumWebdriverSample
{
    // This factory class is just one example of how you might initialize Selenium. If your project already has its own means
    // of initializing Selenium, you can keep using that as-is and ignore this sample file; skip ahead to SamplePageTests.cs.
    public static class WebDriverFactory
    {
        public static IWebDriver CreateFromEnvironmentVariableSettings() {
            // This environment variable gets set by our Azure Pipelines build definition in ./azure-pipelines.yml.
            // That file uses a matrix strategy to run multiple different build jobs for different combinations of OS/browser.
            // Each job sets this environment variable accordingly, and we use it to decide which browser the tests will use.
            const string BROWSER_ENVIRONMENT_VARIABLE = "SELENIUM_BROWSER";
            var browserEnvVar = Environment.GetEnvironmentVariable(BROWSER_ENVIRONMENT_VARIABLE);

            // It's convenient to have a default to make it easier for a developer to run a plain "dotnet test" command.
            // In our CI builds in Azure Pipelines, we'll always specify the browser explicitly instead of using this.
            const string DEFAULT_BROWSER = "chrome";

            switch (browserEnvVar ?? DEFAULT_BROWSER)
            {
                case "chrome":
                    // The ChromeWebDriver environment variable comes pre-set in the Azure Pipelines VM Image we
                    // specify in azure-pipelines.yml. ChromeDriver requires that you use *matching* versions of Chrome and
                    // the Chrome WebDriver; in the build agent VMs, using this environment variable will make sure that we use
                    // the pre-installed version of ChromeDriver that matches the pre-installed version of Chrome.
                    //
                    // See https://docs.microsoft.com/en-us/azure/devops/pipelines/test/continuous-test-selenium
                    //
                    // Environment.CurrentDirectory is where the Selenium.Webdriver.ChromeDriver NuGet package will place the
                    // version of the Chrome WebDriver that it comes with. We fall back to this so that if a developer working on
                    // the project hasn't separately installed ChromeDriver, the test will still be able to run on their machine.
                    var chromeDriverDirectory = Environment.GetEnvironmentVariable("ChromeWebDriver") ?? Environment.CurrentDirectory;

                    // The tests will work fine in non-headless mode; we recommend using --headless for performance and
                    // because it makes it easier to run the tests in non-graphical environments (eg, most Docker containers)
                    var chromeOptions = new ChromeOptions();
                    chromeOptions.AddArgument("--headless");

                    return new ChromeDriver(chromeDriverDirectory, chromeOptions);

                case "firefox":
                    // This environment variable works just like ChromeWebDriver above.
                    var geckoDriverDirectory = Environment.GetEnvironmentVariable("GeckoWebDriver") ?? Environment.CurrentDirectory;
                    
                    // Same as for Chrome above; the tests will work fine without --headless, but you'll get better test
                    // performance and easier compatibility with non-graphical environments by using it.
                    var firefoxOptions = new FirefoxOptions();
                    firefoxOptions.AddArgument("--headless");

                    return new FirefoxDriver(geckoDriverDirectory, firefoxOptions);

                default:
                    throw new ArgumentException($"Unknown browser type '{browserEnvVar}' specified in '{BROWSER_ENVIRONMENT_VARIABLE}' environment variable");
            }
        }
    }
}

File: typescript-selenium-webdriver-sample/tests/webdriver-factory.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as webdriver from 'selenium-webdriver';
import * as chrome from 'selenium-webdriver/chrome';
import * as firefox from 'selenium-webdriver/firefox';
import * as path from 'path';

export function createWebdriverFromEnvironmentVariableSettings(): webdriver.ThenableWebDriver {
    // Selenium WebDriver implementations generally require that you use a version of the webdriver
    // that exactly matches the version of the browser it is driving.
    //
    // Some Azure Pipelines build agents come pre-installed with common browsers and their matching
    // webdriver versions. To detect whether pre-installed webdrivers are available, we check for
    // whether the environment variables ChromeWebDriver and GeckoWebDriver are defined; the
    // vmImages we use in azure-pipelines.yml come with those environment variables pre-set.
    // See https://docs.microsoft.com/en-us/azure/devops/pipelines/test/continuous-test-selenium
    //
    // For local development, where most developers won't have webdrivers pre-installed, we fall
    // back to the chromedriver and geckodriver npm packages, which come bundled with one particular
    // version of the corresponding webdrivers. These packages need to be kept up to date over time
    // with the current stable versions of Chrome/Firefox. This strategy means individual developers
    // on your team won't have to manually keep separate global webdriver installations up to date.
    const azurePipelinesAgentChromeDriverPath = webdriverPathFromEnvVar('ChromeWebDriver', 'chromedriver.exe');
    const chromeDriverPath = azurePipelinesAgentChromeDriverPath || require('chromedriver').path;

    const azurePipelinesAgentGeckoDriverPath = webdriverPathFromEnvVar('GeckoWebDriver', 'geckodriver.exe');
    const geckoDriverPath = azurePipelinesAgentGeckoDriverPath || require('geckodriver').path;

    return new webdriver.Builder()
        // forBrowser sets the *default* browser the tests will use. You can override the defaults
        // by setting the SELENIUM_BROWSER environment variable. This project's package.json includes
        // two npm run scripts ("test:firefox" and "test:chrome") that show examples of using this.
        .forBrowser('chrome')
        .setChromeService(new chrome.ServiceBuilder(chromeDriverPath))
        // You can run accessibility scans on head-ful browsers, too; we recommend using headless
        // browsers unless your project strictly requires head-ful testing, since it will generally
        // be faster, more reliable, and easier to run in non-graphical environments (eg, Docker).
        .setChromeOptions(new chrome.Options().headless())
        .setFirefoxService(new firefox.ServiceBuilder(geckoDriverPath))
        .setFirefoxOptions(new firefox.Options().headless())
        .build();
}

// selenium-webdriver expects paths like "C:\webdrivers-dir\webdriver.exe", but the Azure Pipelines
// environment variables are in form "C:\webdrivers-dir" - this formats them like Selenium expects
function webdriverPathFromEnvVar(directoryEnvironmentVariable: string, binary: string): (string | null) {
    const directory = process.env[directoryEnvironmentVariable];
    return directory ? path.join(directory, binary) : null;
}
